engine/lib/user_settings.php

```diff
@@ -12,333 +12,283 @@
-	$current_password = get_input('current_password', null, false);
-	$password = get_input('password', null, false);
-	$password2 = get_input('password2', null, false);
-	$user_guid = get_input('guid');
-
-	if ($user_guid) {
-		$user = get_user($user_guid);
-	} else {
-		$user = elgg_get_logged_in_user_entity();
+function _elgg_set_user_password(\Elgg\Hook $hook) {
+
+	$actor = elgg_get_logged_in_user_entity();
+
+	$user = $hook->getUserParam();
+	$request = $hook->getParam('request');
+	/* @var $request \Elgg\Request */
+
+	$password = $request->getParam('password', null, false);
+	$password2 = $request->getParam('password2', null, false);
+
+	if (!$password) {
+		return null;
 	}
 
-	if ($user && $password) {
+	if (!$actor->isAdmin() || $user->guid == $actor->guid) {
 		// let admin user change anyone's password without knowing it except his own.
-		if (!elgg_is_admin_logged_in() || elgg_is_admin_logged_in() && $user->guid == elgg_get_logged_in_user_guid()) {
-			$credentials = [
-				'username' => $user->username,
-				'password' => $current_password
-			];
-
-			try {
-				pam_auth_userpass($credentials);
-			} catch (LoginException $e) {
-				register_error(elgg_echo('LoginException:ChangePasswordFailure'));
-				return false;
-			}
-		}
+
+		$current_password = $request->getParam('current_password', null, false);
 
 		try {
-			$result = validate_password($password);
+			elgg()->accounts->assertCurrentPassword($user, $current_password);
 		} catch (RegistrationException $e) {
-			register_error($e->getMessage());
+			$request->validation()->fail('password', '', elgg_echo('LoginException:ChangePasswordFailure'));
+
 			return false;
 		}
+	}
 
-		if ($result) {
-			if ($password == $password2) {
-				$user->setPassword($password);
-				_elgg_services()->persistentLogin->handlePasswordChange($user, elgg_get_logged_in_user_entity());
-
-				if ($user->save()) {
-					system_message(elgg_echo('user:password:success'));
-					return true;
-				} else {
-					register_error(elgg_echo('user:password:fail'));
-				}
-			} else {
-				register_error(elgg_echo('user:password:fail:notsame'));
-			}
-		} else {
-			register_error(elgg_echo('user:password:fail:tooshort'));
-		}
-	} else {
-		// no change
-		return;
+	try {
+		elgg()->accounts->assertValidPassword([$password, $password2]);
+	} catch (RegistrationException $e) {
+		$request->validation()->fail('password', '', $e->getMessage());
+
+		return false;
 	}
 
-	return false;
+	$user->setPassword($password);
+	_elgg_services()->persistentLogin->handlePasswordChange($user, $actor);
+
+	$request->validation()->pass('password', '', elgg_echo('user:password:success'));
 }
 
 /**
  * Set a user's display name
  * Returns null if no change is required or input is not present in the form
  * Returns true or false indicating success or failure if change was needed
  *
- * @return bool|void
+ * @elgg_plugin_hook usersettings:save user
+ *
+ * @param \Elgg\Hook $hook Hook
+ *
+ * @return bool
```

Comment:
why change this?

Reply:
``null`` is not ``void``

Reply:
correct, but `null` also isn't a `bool`

Reply:
https://github.com/phpDocumentor/fig-standards/blob/master/proposed/phpdoc.md#715-return
Let's agree to set it to ``bool|null``. Using ``void`` indicates that the the caller should not expect any return, which is not correct in case of hook handlers.

Reply:
I can agree on `bool|null` but only in cases with more than one return type. So no longer `bool|void`

Reply:
I am documenting it as best practice.